Automattic/sensei
# Read student progress from custom tables

Resolves #7076 

In this PR I did a few important decisions:
1. I use a filter to enable reading from tables. Later, it will be easy to manage the setting hooking into it.
2. I renamed some repositories. As this code is internal, I consider it safe and "legal".

## Proposed Changes
* Add table-reading aggregate repositories for course progress, lesson progress and quiz progress.
* Update repositories' factories to return corresponding implementations depending on enabled features.
* Add `sensei_student_progress_read_from_tables` filter.

## Testing Instructions

1. Enable `tables_based_progress` feature flag: `add_filter( 'sensei_feature_flag_tables_based_progress', '__return_true' );`.
4. Run migrations, wait for the completion. Here, I suppose you have existing progress stored in comments. Otherwise, you need to start and/or complete courses, lessons and quizzes.
5. Add a filter hook: `add_filter( 'sensei_student_progress_read_from_tables', '__return_true' );`
6. Visit and complete courses, lessons and quizzes on the frontend and in the admin UI. There should be no errors.

Unfortunately, in our interfaces, we still use comments-based version of the status. We have a separate card to address this issue: https://github.com/Automattic/sensei/issues/7150

## New/Updated Hooks

* `sensei_student_progress_read_from_tables` - Filter whether to read student progress from tables.

## Pre-Merge Checklist
